zetaclient/chains/base/observer.go

```diff
@@ -290,27 +290,26 @@ func (ob *Observer) StopChannel() chan struct{} {
 
 // OpenDB open sql database in the given path.
 func (ob *Observer) OpenDB(dbPath string, dbName string) error {
-	// create db path if not exist
-	if _, err := os.Stat(dbPath); os.IsNotExist(err) {
-		err := os.MkdirAll(dbPath, os.ModePerm)
-		if err != nil {
-			return errors.Wrapf(err, "error creating db path: %s", dbPath)
+	var dial gorm.Dialector
+
+	// SQLite in-mem db
+	if strings.Contains(dbPath, ":memory:") {
+		dial = sqlite.Open(dbPath)
+	} else {
+		if err := ensureDirectory(dbPath); err != nil {
+			return errors.Wrapf(err, "unable to ensure dbPath %q", dbPath)
 		}
-	}
 
-	// use custom dbName or chain name if not provided
-	if dbName == "" {
-		dbName = ob.chain.ChainName.String()
-	}
-	path := fmt.Sprintf("%s/%s", dbPath, dbName)
+		// use custom dbName or chain name if not provided
+		if dbName == "" {
+			dbName = ob.chain.ChainName.String()
+		}
 
-	// use memory db if specified
-	if strings.Contains(dbPath, ":memory:") {
-		path = dbPath
+		dial = sqlite.Open(fmt.Sprintf("%s/%s", dbPath, dbName))
```

Comment:
nit: Consider moving this statement to a function that standardizes how we expect the db paths.
That way we make this more reusable and deterministic.
```suggestion
		dial = sqlite.Open(buildDBPath(dbPath, dbName))
```